snapcraft/internal/remote_build/_worktree.py

```diff
@@ -14,100 +14,321 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-import fnmatch
+import datetime
+import git
 import logging
 import os
+import shutil
+import subprocess
 
-from pathlib import Path
-from snapcraft.file_utils import link_or_copy, link_or_copy_tree
-from typing import List
-from ._repo import Repo
+from collections import OrderedDict
+from copy import deepcopy
+
+import snapcraft
+import snapcraft.internal.sources
+from snapcraft import yaml_utils
+from snapcraft.project import Project
+from snapcraft.internal.meta import _version
+from snapcraft.internal.remote_build import errors
 
 logger = logging.getLogger(__name__)
 
 
-class Worktree:
-    """Create a copy of the current project directory.
+class WorkTree:
+    """Manages tree for remote-build project."""
+
+    def __init__(self, worktree_dir: str, project: Project) -> None:
+        """Create remote-build WorkTree.
+
+        :param str worktree_dir: Directory to use for working tree.
+        :param Project project: Snapcraft project.
+        """
+        self._project = project
+
+        # Get snapcraft yaml (as OrderedDict).
+        self._snapcraft_config = (
```

Comment:
why store this if you are saving project?
If it is to avoid a second call, you are doing it right below.

Reply:
It's really to avoid several calls to it on anything iterating over / working on parts as the `parts` info is not directly available in project (or .info) AFAICT.  I figured it made sense two grab two copies, one original `self._snapcraft_config` and one working copy to freely modify `self._preopared_snapcraft_config`.

I could replace the refs to `self._snapcraft_config` with `project.__raw_snapcraft`?  I was just trying to avoid use of internal properties.